pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,277 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tagging
+
+import (
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	v1lister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/cloud-provider-aws/pkg/controllers/options"
+	opt "k8s.io/cloud-provider-aws/pkg/controllers/options"
+	awsv1 "k8s.io/cloud-provider-aws/pkg/providers/v1"
+	"k8s.io/klog/v2"
+	"strings"
+	"time"
+)
+
+const (
+	tag string = "tag"
+)
+
+// TaggingController is the controller implementation for tagging cluster resources.
+// It periodically check for Node events (creating/deleting) to apply appropriate
+// tags to resources.
+type TaggingController struct {
+	nodeInformer      coreinformers.NodeInformer
+	controllerOptions options.TaggingControllerOptions
+	kubeClient        clientset.Interface
+	nodeLister        v1lister.NodeLister
+	cloud             *awsv1.Cloud
+	workqueue         workqueue.RateLimitingInterface
+
+	// Value controlling TaggingController monitoring period, i.e. how often does TaggingController
+	// check node list. This value should be lower than nodeMonitorGracePeriod
+	// set in controller-manager
+	nodeMonitorPeriod time.Duration
+
+	// A map presenting the node and whether it currently exists
+	currentNodes map[string]bool
+
+	// A map representing nodes that were ever part of the cluster
+	totalNodes map[string]*v1.Node
```

Comment:
This would need to be garbage collected eventually, why can't we take advantage of the informer to provide us with the node object that has been deleted?

Reply:
See https://github.com/kubernetes/cloud-provider-aws/pull/308#discussion_r835614427